docs/semantics/encapsulation.md

```diff
@@ -0,0 +1,159 @@
+---
+layout: developer-doc
+title: Diagnostics
+category: semantics
+tags: [semantics, diagnostics, runtime]
+order: 11
+---
+
+# Encapsulation
+
+_Encapsulation_ is the system of hiding certain internal **entities** (modules,
+types, methods, constructors, fields) in one project/library from other
+projects/libraries. This document is an excerpt from the specification at
+https://github.com/orgs/enso-org/discussions/7088.
+
+## Requirements
+
+- Be able to hide an entity on demand. By hiding, we mean that the entity cannot
+  be directly imported, and that it cannot be used via FQN.
+  - i.e. an entity shall be hidden both during compile time (project
+    compilation), and during runtime.
+  - Entity being hidden at runtime implies that it does not have any entry in
+    the Suggestion database, therefore, no entry in the Component browser.
+- Be able to import all public symbols from a library with
+  `from Library import all`.
+- Be able to import a selected set of symbols from a library with
+  `from Library import Symbol_1, Symbol_2, ...`.
+- Import a public symbol directly with
+  `import Library.Public_Module.Public_Type`.
+- Use a public symbol via FQN: `Library.Public_Module.Public_Type`.
+
+## Implementation
+
+Let's introduce a `private` keyword. By prepending (syntax rules discussed
+below) `private` keyword` to an entity, we declare it as **project private**. A
+project-private entity is an entity that can be imported and used in the same
+project, but cannot be imported nor used in different projects. Note that it is
+not desirable to declare the entities as _module private_, as that would be too
+restrictive, and would prevent library authors using the entity within the
+project.
+
+From now on, let's consider _project-private_ and _private_ synonymous, and
+**public** as an entity that is not private.
+
+## Syntax
+
+All the entities, except modules, shall be declared private by prepending them
+with `private` keyword. Declaring a module as private shall be done be writing
+the `private` keyword at the very beginning of the module, before all the import
+statements, ignoring all the comments before. Fields cannot have `private`
+keyword, only constructors.
```

Comment:
Can constructors be `private`? Making some constructors `private` and some not will complicate `case of` pattern matching. I was under the impression that one can either hide all constructors or none.